utils/install-app.js

```diff
@@ -139,6 +141,78 @@ function addToServer(appDir, installDir) {
   }
 }
 
+function loadRecognizers(appDir, appId, appVers) {
+  let recognizers;
+  let configRecognizers;
+  let configLocation;
+
+  try { // Get recognizers in a plugin's appDir/config/xxx location
+    recognizers = JSON.parse(fs.readFileSync(path.join(appDir, "config/recognizers", appId))).recognizers;
+    const keys = Object.keys(recognizers)
+    for (const key of keys) { // Add metadata for plugin version & plugin identifier of origin (though objects don't have to be plugin specific)
+      recognizers[key].pluginVersion = appVers;
+      recognizers[key].pluginIdentifier = appId;
+    }
+    logger.debug('ZWED0294I', appId); //logger.debug("Found recognizers for '" + appId + "'");
+  } catch (e) {
+    logger.debug('ZWED0295I', (path.join(appDir, "config/recognizers", appId)), appId); //logger.debug("Could not find recognizers in '" + (path.join(appDir, "config/recognizers", appId)) + "'");
+  }
+
+  try { // Get pre-existing recognizers in config, if any
+    configLocation = path.join(process.env.ZLUX_CONFIG_FILE.substring(1), "../../ZLUX/pluginStorage/org.zowe.zlux.ng2desktop/");
+    configRecognizers = JSON.parse(fs.readFileSync(path.join(configLocation, "recognizers", appId))).recognizers;
+    recognizers = Object.assign(configRecognizers, recognizers); // // If found, combine the ones found in config with ones found in plugin
```

Comment:
If your key for the object (looked like an array, is this correct???) is not meaningfully unique, then object.assign could easily end up with duplicates.

If it's an array, then you need to iterate through and make sure there is no copy based on pluginversion, pluginidentifier, and id.

If it's not an array, then the key could just be pluginid:pluginversion:id

Reply:
I agree, merging algorithm needs to be better